app/src/main/java/org/isoron/uhabits/models/CheckmarkList.java

```diff
@@ -299,4 +303,62 @@ public void writeCSV(Writer out) throws IOException
         cursor.close();
         out.close();
     }
+
+    /**
+     * Writes the checkmark of all habits for each date to the given writer, in CSV format. There is one
+     * line for each date. Each line contains two fields: timestamp and value.
+     *
+     * @param habits the list of habits to write
+     * @param out the writer where the CSV will be output
+     * @throws IOException in case write operations fail
+     */
+
+    public void writeCSVMultipleHabits(List<Habit> habits, Writer out) throws IOException
+    {
+        SimpleDateFormat dateFormat = DateHelper.getCSVDateFormat();
+        String query = "select DISTINCT timestamp from checkmarks order by timestamp";
+        SQLiteDatabase db = Cache.openDatabase();
+        Cursor cursor = db.rawQuery(query, null);
+
+        if(!cursor.moveToFirst()) return;
+
+        out.write(String.format("%s", "Date"));
+        for(Habit habit : habits)
+        {
+            out.write(String.format(",%s", habit.name));
+        }
+        do
+        {
+            String timestamp = dateFormat.format(new Date(cursor.getLong(0)));
+            out.write(String.format("\n%s", timestamp));
+            for(Habit habit : habits)
+            {
+                out.write(String.format(",%s", getHabitValue(habit.getId().toString(), Long.toString(cursor.getLong(0)))));
```

Comment:
Unless I am mistaken, you are making one query per habit per timestamp. If the user has been using the app for one year, and has 50 habits, this will result in almost 20 thousand queries. You can certainly obtain the same information with one query per habit, instead. In fact, you don't even need to write your own queries. Just call the functions `habit.checkmarks.getAllValues()` and `habit.score.getAllValues()`.